File: p279m/num_squares.js
/**
 * @param {number} n
 * @return {number}
 */
var numSquares = function (n) {
  const ans = [0];
  for (let i = 1; i <= n; ++i) {
    let least = Number.MAX_SAFE_INTEGER;
    for (let k = 1; k * k <= i; ++k)
      least = Math.min(least, ans[i - k * k] + 1);
    ans.push(least);
  }
  return ans[n];
};
// TESTS
[
  [1, 1],
  [2, 2],
  [3, 3],
  [4, 1],
  [5, 2],
  [6, 3],
  [7, 4],
  [8, 2],
  [12, 3],
  [13, 2],
].forEach(([n, expected]) => {
  const actual = numSquares(n);
  console.log('The least # of squares sum to', n, '->', actual);
  console.assert(actual === expected);
});

File: p279m/num_squares.py
import math


class Solution:
    def numSquares(self, n: int) -> int:
        squares = [i * i for i in range(1, int(math.sqrt(n)) + 1)]
        dp = [0] * (n + 1)
        for i in range(1, n + 1):
            dp[i] = min(dp[i - k] for k in squares if i - k >= 0) + 1
        return dp[n]


# TESTS
for n, expected in [
    [1, 1],  # 1
    [2, 2],  # 1 + 1
    [3, 3],  # 1 + 1 + 1
    [4, 1],  # 4
    [5, 2],  # 4 + 1
    [6, 3],  # 4 + 1 + 1
    [7, 4],  # 4 + 1 + 1 + 1
    [8, 2],  # 4 + 4
    [12, 3],
    [13, 2],
]:
    sol = Solution()
    actual = sol.numSquares(n)
    print(
        "The least # of perfect squares sum to",
        n,
        "->",
        actual,
    )
    assert actual == expected
